File: src/pay.ts
import { Pay, PAY_REST_API_PROD_URL } from '@binance/pay';
import inquirer from 'inquirer';
import {
    decodeSelectedEntities,
    getConfigurationRestAPI,
    getUserAgent,
    isEmpty,
    readStdinObj,
} from './utils';
import { Parser, hideBin } from 'yargs/helpers';

const parsedArgs = Parser(hideBin(process.argv));

process.env.BINANCE_CONNECTOR_JS_USER_AGENT = getUserAgent('pay');

const stdinObj: any = readStdinObj();

let basePath = PAY_REST_API_PROD_URL;

const configurationRestAPI = getConfigurationRestAPI(parsedArgs?.profile, 'pay');

if (process.env.BINANCE_PAY_BASE_PATH) {
    basePath = process.env.BINANCE_PAY_BASE_PATH;
} else if (configurationRestAPI && configurationRestAPI['basePath']) {
    basePath = configurationRestAPI['basePath'];
}

let client;
if (configurationRestAPI !== null) {
    client = new Pay({
        configurationRestAPI: { ...configurationRestAPI, basePath },
    });
} else {
    client = new Pay({
        configurationRestAPI: {
            apiKey: '',
            basePath,
        },
    });
}

const payCommands: any[] = [];

payCommands.push({
    command: 'get-pay-trade-history',
    describe:
        'Authentication required. ' +
        decodeSelectedEntities(`Get Pay Trade History

* If startTime and endTime are not sent, the recent 90 days&#39; data will be returned.
* The max interval between startTime and endTime is 90 days.
* Support for querying orders within the last 18 months.
* For payerInfo and receiverInfo，there are different return values in different orderTypes.
* Sender&#39;s perspective when orderType is C2C
* payerInfo : binanceId
* receiverInfo : name, binanceId/accountId/email/countryCode/phoneNumber/mobileCode (based on user input)
* Receiver&#39;s perspective when orderType is C2C
* payerInfo : name
* receiverInfo : binanceId
* Sender&#39;s perspective when orderType is CRYPTO_BOX
* payerInfo : binanceId
* receiverInfo : name(the value is always &quot;Crypto Box&quot;)
* Receiver&#39;s perspective when orderType is CRYPTO_BOX
* payerInfo : name
* receiverInfo : binanceId
* Sender&#39;s perspective when orderType is PAY
* payerInfo : binanceId
* receiverInfo : name
* Receiver&#39;s perspective when orderType is PAY
* payerInfo : name
* receiverInfo : binanceId, name
* Sender&#39;s perspective when orderType is PAY_REFUND
* payerInfo : binanceId, name
* receiverInfo : name, accountId
* Receiver&#39;s perspective when orderType is PAY_REFUND
* payerInfo : name
* receiverInfo :  binanceId
* Sender&#39;s perspective when orderType is PAYOUT
* payerInfo : binanceId, name
* receiverInfo : name, accountId
* Receiver&#39;s perspective when orderType is PAYOUT
* payerInfo : name
* receiverInfo :  binanceId
* Receiver&#39;s perspective when orderType is CRYPTO_BOX_RF
* payerInfo : name(the value is always &quot;Crypto Box&quot;)
* receiverInfo : binanceId
* Sender&#39;s perspective when orderType is REMITTANCE
* payerInfo : binanceId
* receiverInfo : name, institutionName, cardNumber, digitalWalletId

Weight: 3000`),
    builder: (yargsCmd: any) => {
        return yargsCmd.options({
            'start-time': {
                describe: decodeSelectedEntities(''),
                type: 'string',
                group: 'Command Options:',
            },
            'end-time': {
                describe: decodeSelectedEntities(''),
                type: 'string',
                group: 'Command Options:',
            },
            limit: {
                describe: decodeSelectedEntities('default 100, max 100'),
                type: 'string',
                group: 'Command Options:',
            },
            'recv-window': {
                describe: decodeSelectedEntities(''),
                type: 'string',
                group: 'Command Options:',
            },
            json: {
                describe: 'Send all fields as JSON',
                type: 'string',
                group: 'JSON Options:',
            },
        });
    },
    handler: async (options: any) => {
        const questions: any = [];

        if (!isEmpty(stdinObj)) {
            options = { ...options, ...stdinObj };
        }

        if (options.json) {
            options = { ...options, ...JSON.parse(options.json) };
            delete options.json;
        }

        if (isEmpty(configurationRestAPI)) {
            console.log(
                'get-pay-trade-history is signed. Please create a profile using `binance-cli profile create`.'
            );
            process.exitCode = 1;
            return;
        }

        if (questions.length > 0) {
            const answers = await inquirer.prompt(questions);
            options = { ...options, ...answers };
        }

        try {
            const response = await client.restAPI.getPayTradeHistory(options);
            const responseData = await response.data();
            console.log(JSON.stringify(responseData, null, 2));
        } catch (e: any) {
            console.log(e.message);
            return;
        }
    },
});

export default {
    command: 'pay',
    description: 'Binance Pay REST API',
    builder: (yargs: any) => {
        yargs.usage('Usage: binance-cli pay <command> [options]');
        payCommands.forEach((command: any) => {
            yargs.command(command);
        });
    },
};

File: src/pay.rs
use crate::utils::{
    build_user_agent, decode_selected_entities, get_configuration_rest_api, read_json_as,
    read_stdin_as,
};
use binance_sdk::config::{ConfigurationRestApi, PrivateKey};
use binance_sdk::constants::PAY_REST_API_PROD_URL;
use binance_sdk::pay::PayRestApi;
use binance_sdk::pay::rest_api::*;
use clap::{Args, Subcommand};
use std::env;
use std::io;
use std::io::{Error, ErrorKind};

fn get_client(profile: Option<&str>, is_signed: bool) -> Result<RestApi, Error> {
    unsafe {
        env::set_var("BINANCE_CONNECTOR_RUST_USER_AGENT", build_user_agent("pay"));
    }

    let config_rest_api = get_configuration_rest_api(profile, "pay").unwrap();
    let api_env = env::var("BINANCE_API_ENV")
        .ok()
        .or(config_rest_api.env)
        .unwrap_or_else(|| "prod".to_string());

    let base_path = match api_env.as_str() {
        "prod" => PAY_REST_API_PROD_URL,
        _ => {
            return Err(Error::new(
                ErrorKind::InvalidInput,
                "invalid BINANCE_API_ENV",
            ));
        }
    };

    let mut builder = ConfigurationRestApi::builder().base_path(base_path);

    if is_signed {
        builder = builder
            .api_key(config_rest_api.api_key)
            .api_secret(config_rest_api.api_secret);

        if config_rest_api.private_key.is_some()  {
            builder = builder.private_key(PrivateKey::File(config_rest_api.private_key.unwrap()));
        }
    }

    let rest_conf = builder
        .build()
        .map_err(|err| Error::new(ErrorKind::Other, err))?;

    Ok(PayRestApi::from_config(rest_conf))
}

#[derive(Args, Debug)]
struct GetPayTradeHistoryArgs {
    #[arg(help = r#"Start time in milliseconds."#, long)]
    start_time: Option<i64>,
    #[arg(help = r#"End time in milliseconds."#, long)]
    end_time: Option<i64>,
    #[arg(help = r#"Number of records to return."#, long)]
    limit: Option<i64>,
    #[arg(help = r#"Request validity window in milliseconds."#, long)]
    recv_window: Option<i64>,
    #[arg(short = 'i', long)]
    interactive: bool,
    #[arg(help = r#"Send all fields as JSON"#, long)]
    json: Option<String>,
    #[arg(help = r#"Select a profile"#, long)]
    profile: Option<String>,
}

#[derive(Subcommand)]
pub enum PayCommands {
    #[command(
        about = decode_selected_entities(r#"Get Pay Trade History

Weight(UID): 3000

Notes:
- If `startTime` and `endTime` are not sent, the recent 90 days' data will be returned.
- The max interval between `startTime` and `endTime` is 90 days.
- Support for querying orders within the last 18 months.
- `payerInfo` and `receiverInfo` return different fields in different `orderType` values:
  - C2C sender: `payerInfo=binanceId`; `receiverInfo=name, binanceId/accountId/email/countryCode/phoneNumber/mobileCode` (based on user input).
  - C2C receiver: `payerInfo=name`; `receiverInfo=binanceId`.
  - CRYPTO_BOX sender: `payerInfo=binanceId`; `receiverInfo=name` (always `"Crypto Box"`).
  - CRYPTO_BOX receiver: `payerInfo=name`; `receiverInfo=binanceId`.
  - PAY sender: `payerInfo=binanceId`; `receiverInfo=name`.
  - PAY receiver: `payerInfo=name`; `receiverInfo=binanceId, name`.
  - PAY_REFUND sender: `payerInfo=binanceId, name`; `receiverInfo=name, accountId`.
  - PAY_REFUND receiver: `payerInfo=name`; `receiverInfo=binanceId`.
  - PAYOUT sender: `payerInfo=binanceId, name`; `receiverInfo=name, accountId`.
  - PAYOUT receiver: `payerInfo=name`; `receiverInfo=binanceId`.
  - CRYPTO_BOX_RF receiver: `payerInfo=name` (always `"Crypto Box"`); `receiverInfo=binanceId`.
  - REMITTANCE sender: `payerInfo=binanceId`; `receiverInfo=name, institutionName, cardNumber, digitalWalletId`."#, false),
    )]
    GetPayTradeHistory(GetPayTradeHistoryArgs),
}

pub async fn handle_pay_command(command: PayCommands) -> anyhow::Result<()> {
    match command {
        PayCommands::GetPayTradeHistory(args) => get_pay_trade_history(args).await,
    }
}

async fn get_pay_trade_history(args: GetPayTradeHistoryArgs) -> anyhow::Result<()> {
    let rest_client = get_client(args.profile.as_deref(), true)?;

    let params = match read_stdin_as::<GetPayTradeHistoryParams>() {
        Some(params) => params,
        None => match args.json {
            Some(json) => read_json_as::<GetPayTradeHistoryParams>(json).ok_or_else(|| {
                io::Error::new(io::ErrorKind::InvalidInput, "failed to parse json params")
            })?,
            None => GetPayTradeHistoryParams::builder()
                .start_time(args.start_time)
                .end_time(args.end_time)
                .limit(args.limit)
                .recv_window(args.recv_window)
                .build()?,
        },
    };

    // Make the API call
    let response = rest_client.get_pay_trade_history(params).await?;

    let data = response.data().await?;
    println!("{}", serde_json::to_string_pretty(&data).unwrap());
    Ok(())
}
